### ISSUE TITLE
build Android common kernel trees with Clang

### ISSUE THREAD

**Issue by nickdesaulniers** [kernelci/kernelci-core#489]
Clang and LLD are **required** for building Android kernels, and there are VTS tests that enforce that.  We should be testing Android trees with Clang (GCC doesn't hurt, but we should also be using the configs that Android uses such as the GKI configs, which may not build with GCC.  I'll file a separate bug about GKI though).  cc @adelva1984 

- [x] #933
- [x] #934
- [x] #935
- [x] #936
- [x] #988

**Comment by musamaanjum** [comment]
I'm starting working on https://github.com/kernelci/kernelci-core/issues/935. We should be aware that build on android tree will fail because of the following two issues in some configurations:
1. https://github.com/ClangBuiltLinux/linux/issues/1473. It has been fixed. The fix is present in linux-next (2e705706), but not in v5.15-rc6 and hence not in Andriod kernels.
2. https://github.com/ClangBuiltLinux/linux/issues/1497. Fixes of these bugs haven't reached to linux-next, mainline and hence not in Andriod kernels.

**Comment by musamaanjum** [comment]
From Nathan on IRC:
> The first patch is in -next and it will be merged into 5.17 when the merge window opens. The second patch is not in -next (but should be) and I was hoping it would be merged in 5.16 but it does not seem like it will be. Once the second patch is at least in -next, clang-14 can be turned on for that tree.

**Comment by gctucker** [comment]
OK, I think we could apply these patches on the staging-next branch for builds run on staging.kernelci.org.  Could you please try to identify these pending patches?

**Comment by musamaanjum** [comment]
1. Fix is in linux-next: 2e70570656ad ("drm/i915: Avoid bitwise vs logical OR warning in snb_wm_latency_quirk()")
2. In [power supply tree](https://git.kernel.org/pub/scm/linux/kernel/git/sre/linux-power-supply.git) which doesn't flow to linux-next: 644106cdb898 ("power: reset: ltc2952: Fix use of floating point literals")

**Comment by gctucker** [comment]
OK thanks, I've added `644106cdb898 ("power: reset: ltc2952: Fix use of floating point literals")` as a patch file for the staging-next kernel branch used on staging.kernelci.org.  It's just copied to a directory on the server where a script is run to push that branch, you'll find it in GitHub when it's run.  We'll see if this fixes some clang-14 builds.

We could create a staging-android branch as well, say based on the 5.15 Android common kernel to apply extra fixes like that and check builds work fine before deploying things in production.  The advantage would be to ensure any build failures are really due to kernel problems and not related to KernelCI (build environment missing something etc.).

@sre I see this patch is on your `fixes` branch but apparently it's not on `for-next`.  Nathan mentioned on IRC that he was expecting the fix to be in linux-next but it's not, as a result.  Is that on purpose?

**Comment by musamaanjum** [comment]
All the sub tickets are complete now. Following is left now:

1. Create `staging-android` branch and add above mentioned 2 patches.
2. Tune the number of android branches that are built with clang-14. Currently, clang has been enabled on 5.10 and 5.15 android branches.

**Comment by musamaanjum** [comment]
Clang-14 builds have been enabled for the following in total now:
* android-mainline
* android-mainline-tracking
* v5.4 variants
* v5.10 variants
* v5.15 variants

**Comment by gctucker** [comment]
@musamaanjum Apart from fine-tuning the number of builds to potentially reduce them, do you see anything left to do before we can close this issue?

**Comment by musamaanjum** [comment]
@gctucker No, there isn't anything left other than that.

**Comment by gctucker** [comment]
OK thanks, I think maintaining the YAML configuration is an on-going task as things keep evolving so this issue can be closed now.
